ZCS-3663 more user friendly SSL error messages for external accounts

tmclane opened the pull request:
Update the code to catch the SSL related certificate exceptions and create a new
IOException that contains a more user-friendly message such that the end user isn't
shown a message that is meaningless. The message shown to the user should empower them
to inform the administrator of their server that their SSL certificate is invalid.

Messages shown to user:
 Encrypted connection: Encrypted connection failed. Please check the SSL certificate on the external server
 STARTTLS: STARTTLS failed. Please check the SSL certificate on the external server.

gordyt commented:
Quick question: I'm wondering if we cannot leave out the first two `catch` clauses for `SSLHandshakeException` as the third one will pick it up in any case.
Note: @tmclane and I talked about this and thought we should get other's opinion.

tmclane commented:
My thinking originally is that our webclient doesn't indicate that STARTTLS WILL BE used in the case of a non-SSL connection being tested.  In other words we are always using encrypted communications to the external server and will therefore validate the certificate.  

Do we want to remove the more specific messages calling out the STARTTLS failure and just report encryption failure?

fatkudu commented on store/src/java/com/zimbra/cs/mailclient/MailConnection.java:
```diff
@@ -80,24 +81,37 @@ public synchronized void connect() throws IOException {
-        } catch (IOException e) {
+        }
+        catch (SSLHandshakeException e) {
+            throw new IOException("Encrypted connection failed. Please check the SSL certificate on the external server.");
```
My view on Gordon's question is that we should only need to catch `SSLHandshakeException` here.
Please log the exception at least at debug level to aid operator trouble shooting if needed.

tmclane replied:
Currently the system is logging the entire stacktrace related to where this IOException originates anyhow.
I don't feel like that adding additional logging is useful at this point unless I can find a way to suppress the stacktrace itself.


Looks like this:

    857534-209:https:https://192.168.56.101/service/soap/TestDataSourceRequest:1516740769414:a9be13b9b025ab62
    E
    ra.common.service.ServiceException.FAILURE(ServiceException.java:288)
    ra.cs.datasource.Pop3Sync.connect(Pop3Sync.java:160)
    ra.cs.datasource.Pop3Sync.test(Pop3Sync.java:112)
    ra.cs.datasource.DataSourceManager.test(DataSourceManager.java:267)
    ra.cs.service.mail.TestDataSource.testDataSource(TestDataSource.java:152)
    ra.cs.service.mail.TestDataSource.handle(TestDataSource.java:52)
    ra.soap.SoapEngine.dispatchRequest(SoapEngine.java:623)
    ra.soap.SoapEngine.dispatch(SoapEngine.java:476)
    ra.soap.SoapEngine.dispatch(SoapEngine.java:274)
    ra.soap.SoapServlet.doWork(SoapServlet.java:303)
    ra.soap.SoapServlet.doPost(SoapServlet.java:213)
    rvlet.http.HttpServlet.service(HttpServlet.java:707)
    ra.cs.servlet.ZimbraServlet.service(ZimbraServlet.java:206)
    rvlet.http.HttpServlet.service(HttpServlet.java:790)
    pse.jetty.servlet.ServletHolder.handle(ServletHolder.java:821)
    pse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1685)
    ra.cs.servlet.CsrfFilter.doFilter(CsrfFilter.java:169)
    pse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1668)
    ra.cs.servlet.RequestStringFilter.doFilter(RequestStringFilter.java:54)
    pse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1668)
    ra.cs.servlet.SetHeaderFilter.doFilter(SetHeaderFilter.java:59)
    pse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1668)
    ra.cs.servlet.ETagHeaderFilter.doFilter(ETagHeaderFilter.java:47)
    pse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1668)
    ra.cs.servlet.ContextPathBasedThreadPoolBalancerFilter.doFilter(ContextPathBasedThreadPoolBalancerFilter.java:107)
    pse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1668)
    ra.cs.servlet.ZimbraQoSFilter.doFilter(ZimbraQoSFilter.java:116)
    pse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1668)
    ra.cs.servlet.ZimbraInvalidLoginFilter.doFilter(ZimbraInvalidLoginFilter.java:117)
    pse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1668)
    pse.jetty.servlets.DoSFilter.doFilterChain(DoSFilter.java:473)
    pse.jetty.servlets.DoSFilter.doFilter(DoSFilter.java:318)
    pse.jetty.servlets.DoSFilter.doFilter(DoSFilter.java:288)
    pse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1668)
    pse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:581)
    pse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
    pse.jetty.security.SecurityHandler.handle(SecurityHandler.java:524)
    pse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:226)
    pse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1158)
    pse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:511)
    pse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:185)
    pse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1090)
    pse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
    pse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:213)
    pse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:109)
    pse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:119)
    pse.jetty.rewrite.handler.RewriteHandler.handle(RewriteHandler.java:318)
    pse.jetty.server.handler.gzip.GzipHandler.handle(GzipHandler.java:437)
    pse.jetty.server.handler.DebugHandler.handle(DebugHandler.java:84)
    pse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:119)
    pse.jetty.server.Server.handle(Server.java:517)
    pse.jetty.server.HttpChannel.handle(HttpChannel.java:306)
    pse.jetty.server.HttpConnection.onFillable(HttpConnection.java:242)
    pse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:261)
    pse.jetty.io.FillInterest.fillable(FillInterest.java:95)
    pse.jetty.io.ssl.SslConnection.onFillable(SslConnection.java:192)
    pse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:261)
    pse.jetty.io.FillInterest.fillable(FillInterest.java:95)
    pse.jetty.io.SelectChannelEndPoint$2.run(SelectChannelEndPoint.java:75)
    pse.jetty.util.thread.strategy.ExecuteProduceConsume.produceAndRun(ExecuteProduceConsume.java:213)
    pse.jetty.util.thread.strategy.ExecuteProduceConsume.run(ExecuteProduceConsume.java:147)
    pse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:654)
    pse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:572)
    g.Thread.run(Thread.java:748)
    IOException: STARTTLS failed. Please check the SSL certificate on the external server
    ra.cs.mailclient.MailConnection.connect(MailConnection.java:99)
    ra.cs.datasource.Pop3Sync.connect(Pop3Sync.java:152)

fatkudu replied:
If I'm reading that correctly, that is logging the ServiceException which has wrapped the new IOException that this pull request creates - thus the original SSLHandshakeException is still lost.  So, I would recommend logging both but perhaps only at debug level.  It can sometimes be useful to log just the message for an exception if the log level is info or above but be more detailed for debug or below.  It kind of looks like the ServiceException in this case doesn't have a useful message - is that possible?  If so, that seems sub-optimal too.

tmclane replied:
Yes that is correct..  The actual exception is a CertificateException but that is actually buried inside the SSLException it seems.  It's like a russian doll of exceptions.

tmclane commented:
Converted exception being thrown to a CertificateException with a message including the Host and Port of the destination that isn't passing validation.
Removed the try/catch surrounding the `startTls` call.

fatkudu commented on store/src/java/com/zimbra/cs/mailclient/MailConnection.java:
```diff
@@ -71,7 +73,7 @@ protected MailConnection(MailConfig config) {
      * @throws IOException if an I/O error occurs
      */
-    public synchronized void connect() throws IOException {
+    public synchronized void connect() throws IOException, CertificateException {
```
This doesn't build.  Callers need modification to cope with CertificationException.

fatkudu commented on store/src/java/com/zimbra/cs/mailclient/MailConnection.java:
```diff
@@ -90,14 +92,21 @@ public synchronized void connect() throws IOException {
-        } catch (IOException e) {
+        }
+        catch (SSLHandshakeException e) {
+            getLogger().debug(String.format("SSL Handshake Exception: %s", e.getMessage()));
```
I think at debug level (which is off by default) it is safe and more useful to log the entire exception.
